packages/flutter/test/rendering/object_test.dart

```diff
@@ -246,6 +246,48 @@ void main() {
     renderObject.dispose();
     expect(renderObject.debugLayer, null);
   });
+
+  test('PaintingContext.pushClip* checks that Clip.none means a null approximate clip', () {
+    int paintCount = 0;
+    void paint(PaintingContext context, Offset offset) {
+      paintCount += 1;
+    }
+
+    final RenderSizedBox child = RenderSizedBox(const Size(300, 300));
+
+    final TestBadClipRenderObject object = TestBadClipRenderObject(child, (PaintingContext context) {
+      paintCount += 1;
+      expect(
+        () => context.pushClipRect(false, Offset.zero, Rect.zero, paint, clipBehavior: Clip.none),
+        throwsFlutterError,
+      );
+      expect(
+        () => context.pushClipRRect(false, Offset.zero, Rect.largest, RRect.zero, paint, clipBehavior: Clip.none),
+        throwsFlutterError,
+      );
+      expect(
+        () => context.pushClipPath(false, Offset.zero, Rect.largest, Path(), paint, clipBehavior: Clip.none),
+        throwsFlutterError,
+      );
+    });
+    layout(object);
+    object.owner!.flushCompositingBits();
+    PaintingContext(ContainerLayer(), Rect.largest).paintChild(object, Offset.zero);
+
+    expect(paintCount, 1);
+  });
+
+  test('PaintingContext.pushClip* allows a RO clip children with none/hardEdge', () {
+    final RenderSizedBox child = RenderSizedBox(const Size(300, 300));
+    final RenderSizedBox specialChild = RenderSizedBox(const Size(200, 200));
+    final TestMultiClipBehaviorClipRenderObject object = TestMultiClipBehaviorClipRenderObject(child, specialChild);
+    layout(object);
```

Comment:
My understanding is that this `layout` call doesn't actually execute the paint method of `TestMultiClipBehaviorClipRenderObject` because the `phase` parameter of `layout` defaults to `EnginePhase.layout`. To execute paint and exercise the code path in `_debugCheckForClipBehavior` that I am concerned about, can you change this line to:

```suggestion
    layout(object, phase: EnginePhase.paint);
```

I would expect that this test would now fail unexpectedly with the FlutterError thrown in _debugCheckForClipBehavior.

Reply:
I'm calling the paint code explicitly in the next line.

Reply:
I've tested this locally and the test still passes - would you prefer I update the test to expect that call returns normally instead of the more explicit call on `PaintingContext.paintChild` below?

Reply:
You are of course right. I miss-read the next lines and paint is executed as expected. Apologies for that....

However, I still think the bug is there. Re-reading this test again, I think it is just hidden by the fact that `TestMultiClipBehaviorClipRenderObject` doesn't implement `visitChildren` correctly (it does not visit the `specialChild`). If you fix that and implement `visitChildren` in `TestMultiClipBehaviorClipRenderObject` correctly (i.e. as shown below) I am pretty sure that this test will fail.

```dart
  @override void visitChildren(RenderObjectVisitor visitor) {
    super.visitChildren(visitor);
    visitor(specialChild);
  }
```

Reply:
Hah oops. I'll take a look

Reply:
Yeah.

I think the two options are: remove the assert, or make it a (new) rule (that probably no one is breaking?) that RenderObjects must clip using the same clip behavior, and if you really want to clip using multiple clipBehaviors you have to create two separate RenderObjects to do it.

That seems like a slightly confusing rule, but it comes with the benefit of having a better error message when you mess things up in the much more common case of incorrectly applying your one and only clip behavior to children, and we could include details about it in the error message.